### DISCUSSION TITLE
can't use graphql

### DISCUSSION THREAD

**Discussion by Samsonov0** [supabase/supabase#11221] [Questions]
Hi! I have been trying to overcome your pg_graph plugin for several weeks, but without success.  I would like to learn more about how to use it on pure js (without frameworks). I found such a guide in the official documentation https://supabase.com/docs/guides/api#graphql-api , but when trying to reproduce what was written, I get 
```
{
"error": "Cannot read properties of null (reading 'useContext')"
}
```
![image](https://user-images.githubusercontent.com/52374669/209286993-7cef81db-577c-435d-a54f-a9746d4aa81d.png)

and the question that interests me no less is: why are the schemes not being created? maybe you need to do something manually?

![image](https://user-images.githubusercontent.com/52374669/209287170-ff14d84d-60f9-4600-82ed-2c6bcda198d8.png)
![image](https://user-images.githubusercontent.com/52374669/209287206-9c8ef410-0fe2-4b5e-8952-3df5a04f5076.png)

**Comment by abhishiv** [comment]
same here, except I get "The schema must be one of the following: public, storage" while trying to access graphql api

**Comment by mhsnook** [comment]
My issue is similar; the graphql schema is empty and I can't create tables on it in the web UI:

![Screenshot from 2022-12-29 00-10-57](https://user-images.githubusercontent.com/1102783/209858035-cd52d8a7-f956-4a39-902c-54ff084ebd3b.png)

I logged in as the postgres user via dbeaver and I am able to do whatever inside the `public` schema, I can't create tables on the `graphql` schema. I keep thinking maybe I am getting something wrong and I need to grant privileges on the `graphql` or `graphql_public` schemas to the anon role... but from what it says in the admin page, I think only the `supabase_admin` role has the rights to do that, and I am not aware of a way for me to log in as that user.

I would be great to have info on these questions:
1. Which schemas can be accessed via the graphql endpoint and do we have any way of controlling this
2. If the answer to (1) is that we can only use the graphql schema, which users have control over table structure and RLS on that schema? Maybe `dashboard_user` and `postgres` roles should have that access? or how can we grant it?
3. Is there any way to have data accessible via both graphql and postgrest endpoints, or for one set of data do we need to choose which on we use and do a hard switch across our whole app?

**Comment by abhishiv** [comment]
just fyi I fixed the issue by creating a new project. That might help in some cases. 

@michaelsnook you can create tables in public schema. That's exposed by graphql API at /graphql/v1?apikey=...

**Reply by mhsnook** [reply]
Ya I also created a new project and I am trying to sort of guess how to access my data this way. I can get the query to work turning the `language` table into `languageCollection`, but I can't seem to find how to access the `card_phrase` table (is it `cardPhraseCollection`? no that doesn't work). Anyhow, it took me _days and days_ to find that this is all documented here: https://supabase.github.io/pg_graphql/ -- I think it's only linked in one place on the entire site??

I don't know if I have a way of actually seeing inside my account what my generated graphql schema is. We have this nice fancy auto-generated API doc for the RESTful API, and that would be cool to have for GraphQL, but even like, just saying what my collections are called, would be a big help. I think my issue on this discussion is mostly resolved, I just wish I knew how to get at the card_phrase table (CardPhraseCollection? cardPhraseCollection? card_phraseCollection??? I tried setting the name myself using a comment on the table, but IDK if that is supposed to take effect immediately or if I have to rebuild the indexes first or if they rebuild on their own at some point?).

**Reply by mhsnook** [reply]
Like, right now, I can access 3 of my public tables
![Screenshot from 2022-12-29 14-35-18](https://user-images.githubusercontent.com/1102783/209929074-4be98f7a-5996-4cf7-a852-6203d5312c0b.png)

But then `card_phrase` and `card_translation` are completely missing? I check the tables in the DB and they are the same -- no RLS 
![Screenshot from 2022-12-29 14-34-43](https://user-images.githubusercontent.com/1102783/209929134-f634a3bb-0f32-435e-826f-7bcf4a3556af.png)

And when I check the table definitions I have the same exact stuff in there for `card_phrase` as for `language`

```sql
GRANT ALL ON TABLE public.card_phrase TO anon;
GRANT ALL ON TABLE public.card_phrase TO authenticated;
GRANT ALL ON TABLE public.card_phrase TO service_role;
...
GRANT ALL ON TABLE public.language TO anon;
GRANT ALL ON TABLE public.language TO authenticated;
GRANT ALL ON TABLE public.language TO service_role;
```

I'm exploring the documentation a bit more and I see under [security](https://supabase.github.io/pg_graphql/security/) it mentions how to expose tables with this example:

```sql
revoke all privileges on public."Account" from api_user;
```

But there is no role `api_user` in my account :sweat_smile: 

So it looks like perhaps that bit of documentation is out of date, or just not specific to the specific setup of pg_graphql installed on my supabase instance. And also it's possible that the GraphQL index hasn't been built completely or properly? but again I am unsure if there's a way for me to inspect those indexes so I am just left guessing. 

I love Supabase in many many ways, but I have been pretty consistently left guessing by the lack of documentation, and just trying to trial graphql has been basically a few days of rabbit-holes and black-holes and time-sucks and because there's very little documentation I'm never sure whether there's a bug in the platform or I'm just doing it wrong :shrug:

**Reply by mhsnook** [reply]
Well I went into the graphiql page and used the network tab to grab the first JSON respose from the API which seems to include the whole schema (yay) but as you can see here there's only 3 collections/tables represented. 

![Screenshot from 2022-12-29 17-13-24](https://user-images.githubusercontent.com/1102783/209946536-dca617bd-2f66-4c2d-94ab-69bb820694ff.png)

So I'm a bit stuck now, and I'm thinking maybe I should instead be looking into this more like a bug with the schema generation parts.

**Reply by mhsnook** [reply]
Okay my issue appears to be resolved. I manually re-ran some grant permissions, and then I created and renamed some tables, and then I went into the admin and noticed that my tables didn't get primary keys set (maybe this was keeping the schema from generating?) and now they are showing up in my Graphiql interface! hooray I think I will go write features and stuff now.

**Reply by Samsonov0** [reply]
> Okay my issue appears to be resolved. I manually re-ran some grant permissions, and then I created and renamed some tables, and then I went into the admin and noticed that my tables didn't get primary keys set (maybe this was keeping the schema from generating?) and now they are showing up in my Graphiql interface! hooray I think I will go write features and stuff now.

hi! can you tell me in more detail what permissions you have restarted? and what kind of admin are we talking about?

**Reply by mhsnook** [reply]
Well, when I first started, I enabled pg_graphql on my old project, but then I thought maybe this is lying to me because the project was created too early, so I did a pg_dump, and then deleted the project (because I had hit my 2-projects limit and at the time couldn't find the option to pause this project) and created a new one and tried to run the import from my postgres backup. 

But I guess I didn't use the right options for the export because it didn't give me all the same setup. And I wasn't sure whether I was supposed to have these tables on the graphql or graphql_public schema (IDK what these do actually; I never figured it out) but I was getting empty responses on some of what should have been public queries, which is common with permissions problems because you don't get 403s you just get empty responses so just to be safe I re-ran some of the things from the original postgres import, statements like these ones:

```sql
GRANT ALL ON TABLE public.language TO anon;
GRANT ALL ON TABLE public.language TO authenticated;
GRANT ALL ON TABLE public.language TO service_role;
```
And then I went into the admin (when I say admin I mean the web interface, app.supabase.com) and (re)created my RLS policies there and set primary keys there (like I said, I must have done the dump/restore incorrectly; I used the verbose flag and some other stuff; my b) and then things worked! At this point I had been floundering for a few days so I wasn't testing in between each step so IDK exactly what caused the problem.

I would guess these are not problems with the Supabase product but probably with how I did my backup/restore. I suppose I would suggest for supabase folks that a) they could document the graphql features a little better (explain that it's the public schema; ignore graphql schemas), and b) maybe don't show the "enable graphql" checkbox for older projects where it won't work. But basically this is probably my error more than anything.
